gufe/protocols/protocolunit.py

```diff
@@ -7,79 +7,379 @@
+
+    @classmethod
+    def _from_dict(cls, dct: Dict):
+        return cls(**dct)
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def source_key(self):
+        return self._source_key
+
+    @property
+    def inputs(self):
+        return self._inputs
+
+    @property
+    def outputs(self):
+        return self._outputs
+
+    @property
+    def dependencies(self) -> List["ProtocolUnitResult"]:
+        """Generate a list of all `ProtocolUnitResult`s dependent on.
+
+        """
+        if self._dependencies is None:
+            self._dependencies = self._list_dependencies(ProtocolUnitResultBase)
+        return self._dependencies     # type: ignore
+
+
+class ProtocolUnitResult(ProtocolUnitResultBase):
+    """Result for a single `ProtocolUnit` execution.
+
+    Attributes
+    ----------
+    name : Optional[str]
+        Name of the `ProtocolUnit` that produced this `ProtocolUnitResult`.
+    source_key : GufeKey
+        Key of the `ProtocolUnit` that produced this `ProtocolUnitResult`
+    inputs : Dict[str, Any]
+        Inputs to the `ProtocolUnit` that produced this
+        `ProtocolUnitResult`. Includes any `ProtocolUnitResult`s this
+        `ProtocolUnitResult` is dependent on.
+    outputs : Dict[str, Any]
+        Outputs from the `ProtocolUnit._execute` that generated this
+        `ProtocolUnitResult`.
+    dependencies : List[ProtocolUnitResult]
+        A list of the `ProtocolUnitResult`s depended upon.
+
+    """
+
+    def ok(self) -> bool:
+        return True
+
+
+class ProtocolUnitFailure(ProtocolUnitResultBase):
+    """Failed result for a single `ProtocolUnit` execution.
+
+    Attributes
+    ----------
+    name : Optional[str]
+        Name of the `ProtocolUnit` that produced this `ProtocolUnitResult`.
+    source_key : GufeKey
+        Key of the `ProtocolUnit` that produced this `ProtocolUnitResult`
+    inputs : Dict[str, Any]
+        Inputs to the `ProtocolUnit` that produced this
+        `ProtocolUnitResult`. Includes any `ProtocolUnitResult`s this
+        `ProtocolUnitResult` is dependent on.
+    outputs : Dict[str, Any]
+        Outputs from the `ProtocolUnit._execute` that generated this
+        `ProtocolUnitResult`.
+    dependencies : List[ProtocolUnitResult]
+        A list of the `ProtocolUnitResult`s depended upon.
+    exception : Exception
+        The exception raised during `ProtocolUnit` exectuion.
+    traceback : str
+        The traceback given by the exception.
+
+    """
+
+    def __init__(self, *, name=None, source_key, inputs, outputs, _key=None, exception, traceback):
+        self._exception = exception
+        self._traceback = traceback
+        super().__init__(name=name, source_key=source_key, inputs=inputs, outputs=outputs, _key=_key)
+
+    def _to_dict(self):
+        dct = super()._to_dict()
+        dct.update({'exception': self.exception,
+                    'traceback': self.traceback})
+        return dct
```

Comment:
RE: Serialization of exceptions/tracebacks: look into [tblib](https://github.com/ionelmc/python-tblib)?

Reply:
would deserialising a traceback require the original dependencies?  i.e. to get an openmm-based traceback would you need openmm installed?  I'm imagining that compute envs might be very different to local envs, and maybe we want results to not require the full stack

Reply:
I'm thinking we might need to go for strings for both exceptions and tracebacks here, given @richardjgowers' point above. If we can come up with something more clever later, then great, but I don't think generally we'll be able to roundtrip exceptions reliably.

Reply:
I've made `ProtocolUnitFailure.exception` give a tuple instead of an exception object. The first element gives the type of exception raised; the second element is a tuple giving the exception's `args` values. This should be serialization/deserialization friendly while still retaining the information we can from the exception (assuming `args` are of simple collections (lists, dicts, etc.) and primitive types).